analyzers/src/SonarAnalyzer.Common/DiagnosticAnalyzer/DiagnosticFactory.cs

```diff
@@ -0,0 +1,31 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers;
+
+public static class DiagnosticFactory
+{
+    public static Diagnostic Create(DiagnosticDescriptor descriptor, Compilation compilation, Location location, IEnumerable<Location> additionalLocations, params object[] messageArgs) =>
+        Diagnostic.Create(descriptor, location, additionalLocations?.Where(x => compilation.ContainsSyntaxTree(x.SourceTree.GetRoot()?.SyntaxTree)), messageArgs);
```

Comment:
What about `location` It could also be in another compilation.

Reply:
Not sure what would be the right thing to do in this case. As we don't have this problem yet I would let this to be handled separately, if/when needed.

Reply:
I think it should be handled in ReportIssue. In DEBUG mode raise an exception and otherwise don't ignore. Do we have a a conditional for peach? If so I would like to do an
```cs
#if peach
  if (DiagnosticContainsInvalidLocation(diagnostic) throw new InvalidOperationException("diagnostic contains invalid location...");
#endif
```

Reply:
We don't have conditionals for peach.